feat: advance sort feature

loxygenK opened the pull request:
つよいソートを実装しました。コンソールでは動いたのでたぶん動きます。MongoDB 上で `loved_user_id` がない古い名言データでも動くのを確認しています!

#### Added
- `--sort=<love|length|id>` ... ソートをします。fav 数・名言の長さ・ID でソートできます。デフォルト値は `id` です。
- `--dir=<a|asc|d|desc>` ... ソート順番を変更します。デフォルトの並び順はソート種類によって異なります。
- `--reversed` (`--rev`, `-R`) … `--dir=desc` です。

#### Breaking change
ないです！最初はあったけど消えた

loxygenK committed:
hopefully it works

kawaemon commented on src/bot/meigen/mod.rs:
```diff
@@ -75,6 +115,23 @@ enum Command {
 
     /// 名言をリスト表示します
+    #[clap(group(
+        ArgGroup::new("sort_conflict")
+            .args(&["sort"])
+            .requires("sort")
+            .conflicts_with("random")
+    ))] // --sort and --random conflicts
```
`--random --sort love` の場合、ランダムにピックして、それらをいいね順にソートして表示　で成立する気がする

loxygenK replied:
する　やる（ジョイマン）

kawaemon commented on src/bot/meigen/mod.rs:
```diff
@@ -75,6 +115,23 @@ enum Command {
+            .conflicts_with("reverse")
+    ))] // --dir and --reverse conflicts
+    #[clap(group(
+        ArgGroup::new("dir_group")
+            .args(&["dir"])
+            .requires_all(["sort", "dir"])
+    ))] // --dir requires --sort
```
`--sort id` がデフォルト値でいいと思う。現在がそうなっているので

loxygenK replied:
まかせろ

kawaemon commented on src/bot/meigen/mod.rs:
```diff
@@ -99,6 +156,20 @@ enum Command {
+        #[clap(value_enum, long)]
+        dir: Option<SortDirection>,
+
+        /// 順番を入れ替えます。
+        #[clap(short, long, alias = "rev")]
+        #[clap(default_value_t = false)]
+        reverse: bool,
```
```suggestion
        /// 降順にします。--dir desc のエイリアスです。
        #[clap(short, long, alias = "rev")]
        #[clap(default_value_t = false)]
        reverse: bool,
```

loxygenK replied:
そうする

kawaemon commented on src/bot/meigen/mod.rs:
```diff
@@ -210,14 +290,11 @@ impl<D: MeigenDatabase> MeigenBot<D> {
             return Ok("条件に合致する名言が見つかりませんでした".into());
         }
 
-        res.sort_unstable_by_key(|x| x.id);
-
```
すまん

loxygenK replied:
#107 に免じてゆるされました

kawaemon commented:
-r と -R 逆にすれば BC にならなくて良いのでは

loxygenK commented:
ただしい　そうする

loxygenK commented:
((trait_ext 赦されたのか))

loxygenK committed:
fix: i was really thinking i formatted but it wasn't (sorry)

loxygenK committed:
fix: consist sort preference / --random works with --sort

loxygenK committed:
feat: --rev is now --sort desc

loxygenK committed:
feat: did all of necessary things for mongodb

loxygenK committed:
fix: i just realized MongoDB sorting logic not correct

loxygenK committed:
refactor: simplification codebase

TakeAndApplyIf wasn't used after all, I'm sobbing so hard but nothing
beats simplicity

loxygenK committed:
feat: `-R` is now `--rev`, and breaking change diminished into void
